server/src/test/java/org/solarbank/server/unit/ServerRestControllerTest.java

```diff
@@ -0,0 +1,49 @@
+package org.solarbank.server.unit;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.solarbank.server.ServerRestController;
+import org.solarbank.server.dto.UserInputDto;
+import org.solarbank.server.service.CalculateService;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.when;
+
+public class ServerRestControllerTest {
```

Comment:
So in a spring application the controller layer is the entry point. So it's customary to write integration tests instead of unit tests at this level.

The idea being that you already have unit tests for all of the steps in the flow, but you don't have tests which validate the flow in its entirety. 

In spring you can use annotations which will spin up your controller and dependency in a test environment and expose the endpoints in that controller for test. Then you can use testing libraries to call the endpoint and assert on the response.

The sort of tests we want to write here would be things like:

- happy path, valid user inputs, returns expected json (exercises your service layer)
- unhappy paths, invalid user inputs returns expected error json (exercises your validation annotations and the global exception handler)

Have a look at section 4 here:

https://www.baeldung.com/spring-boot-testing

Reply:
added integration tests for controller for valid user input and bad request